File: src/main/java/com/qtech/bubbles/common/entity/Modifier.java
package com.qtech.bubbles.common.entity;

import org.bson.BsonDocument;
import org.bson.BsonDouble;
import org.bson.BsonString;

public class Modifier {
    public static final ModifierType ATTACK = new ModifierType("qbubbles.attack");
    public static final ModifierType DEFENSE = new ModifierType("qbubbles.defense");
    public static final ModifierType SCORE = new ModifierType("qbubbles.score");

    private final ModifierType type;
    private final double value;

    public Modifier(ModifierType type, double value) {
        this.type = type;
        this.value = value;
    }

    public double getValue() {
        return value;
    }

    public BsonDocument write(BsonDocument document) {
        document.put("name", new BsonString(this.type.getName()));
        document.put("value", new BsonDouble(this.value));

        return null;
    }

    public static void read(BsonDocument document) {
        ModifierType type = ModifierType.types.get(document.getString("name").getValue());
        double value = document.getDouble("value").getValue();
    }
}

File: src/main/kotlin/qtech/bubbles/common/entity/Modifier.kt
package qtech.bubbles.common.entity

import org.bson.BsonDocument
import org.bson.BsonDouble
import org.bson.BsonString

@Suppress("UNUSED_VARIABLE")
class Modifier(private val type: ModifierType, val value: Double) {
    fun write(document: BsonDocument): BsonDocument? {
        document["name"] = BsonString(type.name)
        document["value"] = BsonDouble(value)
        return null
    }

    companion object {
        val ATTACK = ModifierType("bubbleblaster.attack")
        val DEFENSE = ModifierType("bubbleblaster.defense")
        val SCORE = ModifierType("bubbleblaster.score")
        fun read(document: BsonDocument) {
            val type: ModifierType = ModifierType.Companion.types.get(document.getString("name").value)!!
            val value = document.getDouble("value").value
        }
    }
}